### PR TITLE
Apply moniker to manifest output

### PR DESCRIPTION
**Pull request by live1206**
#3607 
- Add moniker info to `build.manifest`
```json
{
    "files":[
        {
            "siteUrl": "/{SitePath}",
            "outputPath": "{outputPath}",
            "sourcePath": "{sourcePath}",
            "monikers": [
                "moniker1",
                ...
            ]
        },
    ]
}
```
- Add moniker info to legacy `build.manifest`
```json
{
    "groups":[
        "group": "{groupId}",
        "monikers": [
            "moniker1",
            ...
        ],
        ...
    ],
    "files":[
        {
            "siteUrl": "{SitePath}",
            "outputPath": "{outputPath}",
            "sourcePath": "{sourcePath}",
            "group": "{groupId}"
        },
    ]
}
```
> `groupid` is the first 8 characters of the hash of the monikers joined by `,`.

### CODE REVIEW

File: src/docfx/build/Build.cs [hunk 1]
```diff
@@ -166,13 +166,18 @@ void ValidateBookmarks()
                     return DependencyMap.Empty;
                 }
 
+                var (groupId, outputPath) = GetOutputPath(file, monikers);
```
Comment (OsmondJiang) [inline src/docfx/build/Build.cs]: >groupId [](start = 21, length = 7)

what's group?

File: src/docfx/build/Build.cs [hunk 2]
```diff
@@ -166,13 +166,18 @@ void ValidateBookmarks()
...
                 {
                     SourcePath = file.FilePath,
                     SiteUrl = file.SiteUrl,
-                    OutputPath = GetOutputPath(file, monikers),
+                    GroupId = groupId,
+                    OutputPath = outputPath,
+                    File = file,
                 };
 
+                manifest.Monikers.AddRange(monikers);
```
Comment (OsmondJiang) [inline src/docfx/build/Build.cs]: >monikers [](start = 43, length = 8)

why not just assign the value above?

File: src/docfx/build/Build.cs (hunk 1 above)
Comment (live1206) [reply src/docfx/build/Build.cs]: group id is the first 8 characters of the hash of the monikers joined by whitespace.

File: src/docfx/build/Build.cs [hunk 3]
```diff
@@ -201,24 +206,25 @@ void ValidateBookmarks()
             }
         }
 
-        private static string GetOutputPath(Document file, List<string> monikers)
+        private static (string, string) GetOutputPath(Document file, List<string> monikers)
```
Comment (OsmondJiang) [inline src/docfx/build/Build.cs]: > (string, string)  [](start = 22, length = 18)

better to name the every return value

File: src/docfx/build/manifest/FileManifest.cs [hunk 4]
```diff
@@ -10,5 +13,13 @@ internal class FileManifest
         public string OutputPath { get; set; }
 
         public string SourcePath { get; set; }
+
+        public List<string> Monikers { get; } = new List<string>();
+
+        [JsonIgnore]
+        public string GroupId { get; set; }
```
Comment (OsmondJiang) [inline src/docfx/build/manifest/FileManifest.cs]: >GroupId [](start = 22, length = 7)

we are removing group from v3 design

File: src/docfx/build/legacy/manifest/LegacyManifestGroup.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Microsoft.Docs.Build
+{
+    internal class LegacyManifestGroup
+    {
+        [JsonProperty("group")]
```
Comment (OsmondJiang) [inline src/docfx/build/legacy/manifest/LegacyManifestGroup.cs]: >group [](start = 23, length = 5)

is that group mapping to DHS?

**Comment by 928PJY** [comment]
>     > `groupid` is the first 8 characters of the hash of the monikers joined by whitespace.

please correct this

---
Refers to: docs/designs/conceptual-versioning.md:412 in cbb2c98. [](commit_id = cbb2c98dc270492edaaf4f42e2943cda863cc20d, deletion_comment = False)

File: src/docfx/build/Build.cs (hunk 2 above)
Comment (live1206) [reply src/docfx/build/Build.cs]: sure

File: src/docfx/build/legacy/manifest/LegacyManifestGroup.cs (hunk 5 above)
Comment (live1206) [reply src/docfx/build/legacy/manifest/LegacyManifestGroup.cs]: yes

File: src/docfx/build/manifest/FileManifest.cs (hunk 4 above)
Comment (live1206) [reply src/docfx/build/manifest/FileManifest.cs]: renamed, just the same value

File: src/docfx/build/manifest/FileManifest.cs [hunk 6]
```diff
@@ -10,5 +13,13 @@ internal class FileManifest
         public string OutputPath { get; set; }
 
         public string SourcePath { get; set; }
+
+        public List<string> Monikers { get; set; }
+
+        [JsonIgnore]
+        public string MonikerSeg { get; set; }
```
Comment (OsmondJiang) [inline src/docfx/build/manifest/FileManifest.cs]: >MonikerSeg [](start = 22, length = 10)

add summary to this property

Comment (yufeih) [reply src/docfx/build/manifest/FileManifest.cs]: this can be computed in legacy.

File: src/docfx/build/manifest/FileManifest.cs [hunk 7]
```diff
@@ -10,5 +13,13 @@ internal class FileManifest
...
+
+        [JsonIgnore]
+        public Document File { get; set; }
```
Comment (yufeih) [inline src/docfx/build/manifest/FileManifest.cs]: can we remove `Document File` from this class, I want our data contract to be plain old c# class without [JsonIgnore], there are other ways to pass objects around without injecting objects to data contract.

**Pull request merged**